### ISSUE TITLE
inheritance error crsfml library

### ISSUE THREAD

**Issue by ekicam2** [crystal-lang/crystal#2293]
error while compiling:

```
Bug: `@owned` at /home/ekicam2/Projects/Crystal/crsfml/graphics_obj.cr:208:44 has no type
[5670246] ???
[9373650] ???
[9283579] ???
[9366822] ???
[9289281] ???
[9356734] ???
[9348776] ???
[9387319] ???
[9278503] ???
[9335578] ???
[9291148] ???
[9397091] ???
[9384641] ???
[9278503] ???
[9279628] ???
[9356734] ???
[9348776] ???
[9387319] ???
[9278503] ???
[9279628] ???
[9356734] ???
[9348776] ???
[9387319] ???
[9278503] ???
[9364684] ???
[9280354] ???
[7986380] ???
[12564579] ???
[12377920] ???
[12372667] main +23387
[140301958092560] __libc_start_main +240
[5668313] _start +41
[0] ???
Error: you've found a bug in the Crystal compiler. Please open an issue, including source code that will allow us to reproduce the bug: https://github.com/manastech/crystal/issues
```

hello.cr:

```
window = SF::RenderWindow.new(SF.video_mode(800, 600), "CrSFML works!")
myShape = MyClass.new

while window.open?
    while event = window.poll_event()
        if event.type == SF::Event::Closed
            window.close()
        end
    end

    window.clear(SF::Color::Black)

    window.draw(myShape)

    window.display()
end
```

myClass.cr:

```
require "./crsfml/graphics"

class MyClass < SF::CircleShape
    def initialize
        super(50)
        self.fill_color = SF.color(150, 50, 250)
    end
end
```

this bug occurs when calling super method in MyClass if we coment this line there will be normal errors hope it will help you guys! Good luck.

**Comment by vjdhama** [comment]
@ekicam2 IMO, you should probably file this bug in the [crsfml](https://github.com/BlaXpirit/crsfml) project. cc/ @BlaXpirit

**Comment by BlaXpirit** [comment]
Changing this to `def initialize(asdf : Int)` (adding any random argument) makes it work.

The reason for failure: [CircleShape](https://github.com/BlaXpirit/crsfml/blob/420d2cb687b14f5e5bbb82c30d587a293959d9ea/src/graphics.cr#L173).initialize(args) calls the `initialize` without arguments. Because you overrode this `initialize`, it seems to want to call your version instead, causing infinite recursion.
I'm not sure what the right behavior of Crystal would be; how to avoid this problem.

**Comment by asterite** [comment]
Interesting. I'll try to take a look at this, it probably makes sense to search the `initialize` method from the given type and up the hierarchy. Thanks for the explanation, @BlaXpirit :-)

**Comment by jhass** [comment]
Got this for triage, could still reproduce this on 0.18. `super` in general should only search from the current type upwards, no matter in whatever context. Everything else is rather unintuitive.

**Comment by j8r** [comment]
Reproducible example:
```cr
class Foo
  def initialize
  end

  def initialize(i)
    initialize
  end
end

class Bar < Foo
  def initialize
    super(1)
  end
end

Bar.new
```

**Comment by straight-shoota** [comment]
The same behaviour reproduces with regular methods, not just `initialize`. And it does not depend on using `super`. For more simplicity I'm using different names instead of arity to express different methods in the following examples.

It generally makes sense to select the implementation in a subclass in cases like this one:
```cr
class Foo
  def foo
    "Foo#foo"
  end

  def bar
    foo
  end
end

class Bar < Foo
  def foo
    "Bar#foo"
  end
end

Bar.new.bar # => "Bar#foo"
```

A problem arises when the call to `#foo` in the superclass is fine and expected, but in a subclass it ends up creating a feedback loop:
```cr
class Foo
  def foo
    "Foo#foo"
  end

  def bar
    foo
  end
end

class Bar < Foo
  def foo
    bar
  end
end

typeof(Bar.new.foo) # => NoReturn
```

The compiler technically recognizes that this doesn't end well, hence the method call is typed as `NoReturn`.

But I think this is not an issue the compiler should be able to fix. It's a simple code error: two methods are calling each other again and again. In the OP it's just a bit hidden by the complexity of inheritance.

Maybe the compiler could offer help to discover and consequently avoid such a problem. But it can't magically fix it. The code needs to be changed, not the code's interpretation.

Changing the code is not trivial, though. It requires changing the implementation of `#bar` to delegate to an internal method with the same behaviour as `#foo`, for example. Alternatively, the signature of the child type's method could be changed to not override the parent definition (at least not for the affected call).

Loosely related: https://forum.crystal-lang.org/t/call-a-method-on-parent-thats-also-defined-on-the-child-from-a-different-method/3493

**Comment by asterite** [comment]
> The same behaviour reproduces with regular methods, not just initialize

@straight-shoota Yes, but that's fine. With `initialize` it's wrong because when you construct an object you always want to construct that object, not a parent type.

But I need more time to fully understand this. But my guess is, if you try this in Java or some other OOP language, constructors are never virtual and they are never looked below the current type.

**Comment by straight-shoota** [comment]
Yeah, maybe constructors should have different rules.

In Java this is a bit different because constructors are named after the type they construct, while in Crystal all constructors/initializers are called `#initialize`.

The equivalent in Java would be this:
```java
public class HelloWorld{
  static class Foo {
    public Foo() {
    }
  
    public Foo(int i) {
      this();
    }
  }
  
  static class Bar extends Foo {
    public Bar() {
      super(1);
    }
  }
  
  public static void main(String []args){
    System.out.println(new Bar());
  }
}
```
And it works as expected, `this()` resolves to calling `Foo()` constructor, not `Bar()`.

Ruby on the other hand has the same infinite loop issue as Crystal (the example from https://github.com/crystal-lang/crystal/issues/2293#issuecomment-655690106 works as literal Ruby code).

**Comment by asterite** [comment]
I think in Ruby it doesn't count because Ruby doesn't have overloads.
